### PR TITLE
Add model YAML serialisation

### PR DESCRIPTION
**Pull request by QRemy**
read/write models from/to yaml
add import/export functions in utils.serialization.io
add .to_yaml()  to SkyModels
use oyaml module to preserve dict order in yaml

### CODE REVIEW

File: gammapy/utils/fitting/parameter.py [hunk 1]
```diff
@@ -354,10 +362,10 @@ def from_dict(cls, val):
                 Parameter(
                     name=par["name"],
                     factor=float(par["value"]),
-                    unit=par["unit"],
-                    min=float(par["min"]),
-                    max=float(par["max"]),
-                    frozen=par["frozen"],
+                    unit=par["unit"] if "unit" in par.keys() else '',
```
Comment (cdeil) [inline gammapy/utils/fitting/parameter.py]: Is this a dict? Then you can use `par.get("unit", "")` pattern for defaults.

File: gammapy/utils/scripts.py [hunk 2]
```diff
@@ -2,7 +2,7 @@
 """Utils to create scripts and command-line tools"""
 import sys
 import logging
-import yaml
+import oyaml as yaml
```
Comment (cdeil) [inline gammapy/utils/scripts.py]: @QRemy - Can you use `pyyaml` for now?

Recently there was a lot of discussion which YAML package to use, and the conclusion was to use `pyyaml`, see:
https://github.com/gammapy/gammapy/blob/321cb278c6f1f507fb322468f70fa02c8d9d41d4/docs/development/pigs/pig-013.rst#alternatives

Do you absolutely need ordered dict for now? If you only read YAML and don't write YAML text and assert on it, then undefined order shouldn't matter for now, no?

File: gammapy/utils/serialization/io.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
```
Comment (cdeil) [inline gammapy/utils/serialization/io.py]: @QRemy - Please run the `black` code formatter on all the files you edit here.

File: gammapy/utils/scripts.py (hunk 2 above)
Comment (QRemy) [reply gammapy/utils/scripts.py]: Ordered yaml files would be more human readable.
oyaml is based on pyyaml it just adds a  representer to enforce order when writing the file:
https://github.com/wimglenn/oyaml/blob/master/oyaml.py
if you don't want to import this module we can just copy the code to utils/scripts, no ?

File: gammapy/utils/fitting/parameter.py (hunk 1 above)
Comment (QRemy) [reply gammapy/utils/fitting/parameter.py]: Yes

File: gammapy/utils/fitting/parameter.py [hunk 4]
```diff
@@ -179,17 +179,25 @@ def __repr__(self):
...
-            name=self.name,
-            value=self.value,
-            factor=self.factor,
-            scale=self.scale,
-            unit=self.unit.to_string("fits"),
-            min=self.min,
-            max=self.max,
-            frozen=self.frozen,
-        )
+    def to_dict(self, do_cut=False):
```
Comment (cdeil) [inline gammapy/utils/fitting/parameter.py]: Suggest to change `do_cut` to this:
```
Parameters
-----------
selection : {"all", "simple"}
    Selection of information to include.
```

To me "do_cut" is not intuitive, prefer "selection" over "cut".

Making this a multi-choice string option means we can add other subsets in the future while keeping backwards-compatibility, whereas if we put a bool option we'll break every caller if we want another subset in the future. Not sure if we really need this flexibility, but I would go with the selection string option.

File: gammapy/utils/scripts.py (hunk 2 above)
Comment (cdeil) [reply gammapy/utils/scripts.py]: Suggest to try using `sort_keys=False` with pyyaml
```
>>> import yaml
>>> yaml.dump({2:2, 1:1}, sort_keys=False)
'2: 2\n1: 1\n'
```
Added very recently: https://github.com/yaml/pyyaml/blob/5986257f9fc978d4a61b6e0001df554f80e565cb/CHANGES#L33

If there's CI fails, I can help to fix them up by adjusting pyyaml versions.

File: gammapy/utils/scripts.py [hunk 5]
```diff
@@ -91,6 +91,26 @@ def make_path(path):
     # Otherwise this can result in cryptic errors later on
     return Path(expandvars(str(path)))
 
+def name_from_path(filepath):
```
Comment (cdeil) [inline gammapy/utils/scripts.py]: What does `name_from_path` do? It's never called?

This shouldn't be needed, we use `pathlib.Path` and that has a lot of convenience methods.

Please delete.
Or show a caller & test & use case, and we can try to make it simpler.

File: gammapy/utils/serialization/tests/test_io.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+from ....spectrum import models as spectral
+from ....image import models as spatial
+from ....cube.models import SkyModels
+from ...scripts import read_yaml, write_yaml
+from ...serialization import models_to_dict, dict_to_models
+from astropy.utils.data import get_pkg_data_filename
+
+@requires_data()
+def test_yaml_io_3d():
+#    filename = "$GAMMAPY_DATA/tests/models/examples.yaml"
```
Comment (cdeil) [inline gammapy/utils/serialization/tests/test_io.py]: Remove comment.

**Comment by cdeil** [comment]
I've done some edits in c277c3411 .

With the current code, there were circular import errors when running `pytest gammapy` between io.py and xml.py and the file where the model classes are. This is now fixed by putting a delayed import in the model files `to_yaml` methods, following what we had for the `to_xml` methods.

I'm not sure if that is the right direction to go, or if we should put the opposite. @adonath @QRemy - Thoughts?

I also have a strong preference for as much as possible lower-case keys. I changed the code, but the example file has a lot of upper-case keys. @QRemy - Is there a way to generate the example data file? Do you have a script? Share it, e.g. as `make_example_data.py` in `gammapy/utils/serialisation/tests`?

**Comment by cdeil** [comment]
Concerning the question whether `serialize.py` or `models.py` should have delayed imports, I can see arguments either way.

But it seems importing the data classes in a `serialize.py` at the top is very common:
https://hynek.me/articles/serialization/
https://github.com/astropy/astropy/blob/94cdb1be51d73e7ee6e1c188d7132f33f6cccfcc/astropy/io/misc/yaml.py#L73

**Comment by cdeil** [comment]
I think we should merge now, if CI passes.

There's several things that need to be changed, but as discussed previously we'll do that as a series of PRs and then have a larger review at the end, instead of a long-running 1000+ line of code feature branch.

**Comment by cdeil** [comment]
This fails because YAML sort_keys was only added very recently:
https://travis-ci.org/gammapy/gammapy/jobs/563981180#L1360

I'm removing that argument now in fc641e614 so that this PR can go in.
We will get sorted keys.
The tests don't rely on order currently.

I think we should change to `sort_keys=False`, but do it later once #2218 is approved,
I'll mention it as a reason to adopt modern pyyaml there.
https://github.com/yaml/pyyaml/blob/5986257f9fc978d4a61b6e0001df554f80e565cb/CHANGES#L33

**Comment by cdeil** [comment]
@QRemy - Thanks!

**Comment by adonath** [comment]
Thanks @QRemy and @cdeil! I'll go ahead an prepare the release now...

**Pull request merged**